### PR TITLE
Structuring dependencies

### PR DESCRIPTION
**Pull request by petersteneteg**
Quite a bit of refactoring, trying to always use the pattern
find_package(* REQUIRED) 
and then 
target_link_libraries(sgct PUBLIC *)
for all dependencies and guarding all included dependencies behind SGCT_DEP_INCLUDE_*

Have not tested all the apps, seems some compile errors in OpenVR, was not able to test OpenAL
and with ffmpeg I'm getting lots of link issues...

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -9,50 +9,68 @@
 cmake_minimum_required(VERSION 3.10 FATAL_ERROR)
 project(sgct)
 
-include(support/cmake/copy_files.cmake)
-include(support/cmake/copy_sgct_dynamic_libraries.cmake)
-include(support/cmake/set_compile_options.cmake)
-include(support/cmake/disable_external_warnings.cmake)
-
-set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${PROJECT_SOURCE_DIR}/support/cmake/modules)
+if(POLICY CMP0056)
```
Comment (alexanderbock) [inline CMakeLists.txt:12]: Coding style
We have spaces after `if` in CMake

File: CMakeLists.txt [hunk 2]
```diff
@@ -9,50 +9,68 @@
...
 project(sgct)
 
-include(support/cmake/copy_files.cmake)
-include(support/cmake/copy_sgct_dynamic_libraries.cmake)
-include(support/cmake/set_compile_options.cmake)
-include(support/cmake/disable_external_warnings.cmake)
-
-set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${PROJECT_SOURCE_DIR}/support/cmake/modules)
+if(POLICY CMP0056)
+     cmake_policy(SET CMP0056 NEW)
```
Comment (alexanderbock) [inline CMakeLists.txt:13]: Coding style
We use 2 space increment for CMake files

File: CMakeLists.txt [hunk 3]
```diff
@@ -9,50 +9,68 @@
...
+
+include(support/cmake/copy_files.cmake)
+include(support/cmake/copy_sgct_dynamic_libraries.cmake)
+include(support/cmake/set_compile_options.cmake)
+include(support/cmake/disable_external_warnings.cmake)
+
+include(support/cmake/register_package.cmake)
+file(REMOVE_RECURSE ${PROJECT_BINARY_DIR}/pkg)
+list(APPEND CMAKE_MODULE_PATH 
+	${PROJECT_SOURCE_DIR}/support/cmake/modules
```
Comment (alexanderbock) [inline CMakeLists.txt:35]: Coding style
See above

File: ext/CMakeLists.txt [hunk 4]
```diff
@@ -6,52 +6,82 @@
 # For conditions of distribution and use, see copyright notice in LICENSE.md             #
 ##########################################################################################
 
-add_subdirectory(glad)
-set_property(TARGET glad PROPERTY FOLDER "Dependencies")
+if(SGCT_DEP_INCLUDE_GLAD)
```
Comment (alexanderbock) [inline ext/CMakeLists.txt:9]: Coding Style
See above

File: ext/CMakeLists.txt [hunk 5]
```diff
@@ -6,52 +6,82 @@
...
 ##########################################################################################
 
-add_subdirectory(glad)
-set_property(TARGET glad PROPERTY FOLDER "Dependencies")
+if(SGCT_DEP_INCLUDE_GLAD) 
+  add_subdirectory(glad)
+  set_property(TARGET glad PROPERTY FOLDER "Dependencies")
+  disable_external_warnings(glad)
+  sgct_register_package(glad glad::glad)
+endif()
```
Comment (alexanderbock) [inline ext/CMakeLists.txt:14]: Coding style
see above

File: ext/glad/CMakeLists.txt [hunk 6]
```diff
@@ -11,4 +11,5 @@ if (WIN32)
   set(FILES ${FILES} "src/glad_wgl.c")
 endif ()
 add_library(glad STATIC ${FILES})
-target_include_directories(glad PUBLIC include)
+add_library(glad::glad ALIAS glad)
+target_include_directories(glad PUBLIC include)
```
Comment (alexanderbock) [inline ext/glad/CMakeLists.txt:15]: Missing empty line at the end of the file

File: src/apps/simplenavigation/main.cpp [hunk 7]
```diff
@@ -394,8 +394,8 @@ std::vector<std::byte> encode() {
     return data;
 }
 
-void decode(const std::vector<std::byte>& data, unsigned int pos) {
-    deserializeObject(data, pos, xform);
+void decode(const std::vector<std::byte>& data, unsigned int apos) {
```
Comment (alexanderbock) [inline src/apps/simplenavigation/main.cpp:397]: `apos`?

File: src/sgct/CMakeLists.txt [hunk 8]
```diff
@@ -134,127 +134,104 @@ set(SOURCE_FILES
   utils/sphere.cpp
 )
 
-if (SGCT_OPENVR_SUPPORT)
-  set(HEADER_FILES ${HEADER_FILES} ${PROJECT_SOURCE_DIR}/include/sgct/openvr.h)
-  set(SOURCE_FILES ${SOURCE_FILES} openvr.cpp)
-endif ()
+if(SGCT_OPENVR_SUPPORT)
```
Comment (alexanderbock) [inline src/sgct/CMakeLists.txt:137]: Coding style
see above

File: src/sgct/CMakeLists.txt [hunk 9]
```diff
@@ -134,127 +134,104 @@ set(SOURCE_FILES
...
+find_package(glfw3 REQUIRED)
+find_package(PNG REQUIRED)
+find_package(tinyxml2 REQUIRED)
+find_package(OpenGL REQUIRED)
 
-target_link_libraries(sgct PUBLIC tracy)
-if (SGCT_DEP_ENABLE_TRACY)
-  target_compile_definitions(sgct PUBLIC SGCT_HAS_TRACY TRACY_ENABLE)
-endif ()
+if(SGCT_VRPN_SUPPORT)
```
Comment (alexanderbock) [inline src/sgct/CMakeLists.txt:170]: Coding style
see above

File: support/cmake/register_package.cmake [hunk 10]
```diff
@@ -0,0 +1,15 @@
+set(localModuleDir ${PROJECT_BINARY_DIR}/pkg)
+
+function(sgct_register_package name)
+    set(incdirs "")
+    foreach(target ${ARGN})
```
Comment (alexanderbock) [inline support/cmake/register_package.cmake:5]: Coding style
see above

File: support/cmake/register_package.cmake [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+        get_target_property(dirs ${target} INTERFACE_INCLUDE_DIRECTORIES)
+        set(incdirs "${incdirs} ${dirs}")
+    endforeach()
+    file(WRITE "${localModuleDir}/Find${name}.cmake" 
+         "# Fake Find file for ${name}\n"
+         "set(${name}_FOUND ON)\n"
+         "set(${name}_LIBRARIES ${ARGN})\n"
+         "set(${name}_INCLUDE_DIRS ${incdirs})\n"
+         )
+ endfunction()
```
Comment (alexanderbock) [inline support/cmake/register_package.cmake:15]: Coding style
Missing empty line at the end of the file

File: CMakeLists.txt (hunk 1 above)
Comment (petersteneteg) [reply CMakeLists.txt:12]: ugly but fair enough ;) don't think you used it consistently though...

File: CMakeLists.txt (hunk 2 above)
Comment (petersteneteg) [reply CMakeLists.txt:13]: ahh right...

File: src/apps/simplenavigation/main.cpp (hunk 7 above)
Comment (petersteneteg) [reply src/apps/simplenavigation/main.cpp:397]: you have a global pos above, hence I get a shadow warning here...

**Comment by petersteneteg** [comment]
@alexanderbock should be fixed now

**Pull request merged**